Sources/Purchasing/StoreKit2/Observer Mode/StoreKit2ObserverModePurchaseDetector.swift

```diff
@@ -0,0 +1,124 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreKit2ObserverModePurchaseDetector.swift
+//
+//  Created by Will Taylor on 5/1/24.
+
+import Foundation
+import StoreKit
+
+/// A delegate protocol for handling verified transactions in observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+// swiftlint:disable type_name
+protocol StoreKit2ObserverModePurchaseDetectorDelegate: AnyObject, Sendable {
+
+    /// Handles a verified transaction with its corresponding JWS representation.
+    /// - Parameters:
+    ///   - verifiedTransaction: The verified transaction to be processed.
+    ///   - jwsRepresentation: The JSON Web Signature representation of the transaction.
+    func handleSK2ObserverModeTransaction(
+        verifiedTransaction: StoreKit.Transaction,
+        jwsRepresentation: String
+    ) async throws
+}
+
+/// Protocol describing an actor capable of detecting purchases from StoreKit 2.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+protocol StoreKit2ObserverModePurchaseDetectorType: Actor {
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModePurchaseDetectorDelegate?
+    ) async
+}
+
+/// Actor responsibile for detecting purchases from StoreKit2 that should be processed by observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+actor StoreKit2ObserverModePurchaseDetector: StoreKit2ObserverModePurchaseDetectorType {
+
+    private let deviceCache: DeviceCache
+    private let currentUserProvider: CurrentUserProvider
+    private let allTransactionsProvider: AllTransactionsProviderType
+
+    init(
+        deviceCache: DeviceCache,
+        currentUserProvider: CurrentUserProvider,
+        allTransactionsProvider: AllTransactionsProviderType
+    ) {
+        self.deviceCache = deviceCache
+        self.currentUserProvider = currentUserProvider
+        self.allTransactionsProvider = allTransactionsProvider
+    }
+
+    /// Detects unobserved transactions and forwards them to the StoreKit2ObserverModeManagerDelegate
+    /// for processing.
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModePurchaseDetectorDelegate?
```

Comment:
if we're expecting it to never be `nil`, we should maybe look for a safer approach

Reply:
I've made the delegate a non-optional param here! If we want to simplify this further, this could be made into a callback param, but I think it's probably fine the way it is.